### ISSUE TITLE
Expose `AuthorizationError`

### ISSUE THREAD

**Issue by khieta** [cedar-policy/cedar#114]
### Category

User level API changes

### Describe the feature you'd like to request

`EvaluationError` is currently exposed through `cedar/api.rs` as a string message (code copied below). This may encourage users to pattern match on our error messages, which are subject to change. We should consider updating this to match cedar-policy-core's `EvaluationError` (per the TODO), or construct some new enum. 
```
/// Errors encountered while evaluating policies or expressions, or making
/// authorization decisions.
#[derive(Debug, Clone, PartialEq, Eq, Error)]
pub enum EvaluationError {
    /// Error message, as string.
    /// TODO in the future this can/should be the actual Core `EvaluationError`
    #[error("{0}")]
    StringMessage(String),
}
```

### Describe the solution you'd like

See above.

### Describe alternatives you've considered

N/A

### Is this something that you'd be interested in working on?

- [x] 👋 I may be able to implement this feature request
- [x] ⚠️ This feature might incur a breaking change

**Comment by NetwCodeProjects** [comment]
## Update to `is_authorized` Function
The function is defined on lines 93-159 in the `cedar-policy-core/src/authorizer.rs`.
It needs to be updated to return error information using a `Vec<EvaluationError>` instead of `Vec<String>`.
### Solution
I'm going to break down some of the changes that need to be made so reviewers can follow along with the logic.
- Update the function signature to include the return type `Vec<EvaluationError>`

   ```rust
   pub fn is_authorized(&self, q: &Request, pset: &PolicySet, entities: &Entities) -> Vec<EvaluationError> {
       // Function implementation
   }
  ```
- Change the declaration of errors to be a `Vec<EvaluationError>` instead of `Vec<String>`
  ```rust
  let mut errors: Vec<EvaluationError> = partial.diagnostics.errors;
  ```
- The part that extends the errors vector needs to be modified to construct instances of EvaluationError instead of using `format!` to create strings.
  - Check EvaluationError::NonValue is the right constructor for the EvaluationError type in the `cedar-policy-core`. 
  ```rust
  errors.extend(partial.residuals.policies().map(|p| {
    EvaluationError::NonValue(p.condition())
  }));
- Update the code, to create the response—using the modified errors vector.
  ```rust
  match self.error_handling {
    ErrorHandling::Deny => Response::new(
        Decision::Deny,
        idset.chain(partial.diagnostics.reason.into_iter()).collect(),
        errors,
    ),
    ErrorHandling::Forbid => Response::new(
        Decision::Deny,
        idset.chain(partial.diagnostics.reason.into_iter()).collect(),
        errors,
    ),
    ErrorHandling::Skip => {
        // Logic to determine the response based on errors and residuals
        if partial.residuals.policies().any(|p| p.effect() == Effect::Permit) {
            Response::new(
                Decision::Allow,
                partial.diagnostics.reason.into_iter().chain(idset).collect(),
                errors,
            )
        } else {
            Response::new(
                Decision::Deny,
                partial.diagnostics.reason.into_iter().chain(idset).collect(),
                errors,
              )
          }
      }
  }
  ```
Determine the response based on errors and residuals.
The `if statement` checks if any residual policies have an effect of Permit. If there are such policies, the response is created with `Decision::Allow`, combining partial.diagnostics.reason, idset, and errors. 
Otherwise, the response is created with `Decision::Deny`, combining the same elements.

The key differences are:
1. The original code checks if any policy in partial.diagnostics.reason has an effect of Permit. In contrast, the updated code checks if any policy in partial.residuals.policies() has an effect of Permit.

2. The original code uses partial.diagnostics.reason directly in the response creation, while the updated code uses partial.diagnostics.reason.into_iter().chain(idset) to combine the reasons from partial.diagnostics and idset.

Overall, the updated code aims to determine the response based on the modified errors vector and the presence of permit policies in the residuals, while the original code focuses on the presence of permit policies in `partial.diagnostics.reason`.

Consult the comments on line 127 for further understanding.
## Further changes
There are a few `Vec<string>` calls that need to be updated.
For example, in the Diagnostics struct.
- We need to update the Diagnostics struct to use `Vec<EvaluationError>` for the errors field instead of `Vec<String>`
```rust
#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
pub struct Diagnostics {
    pub reason: HashSet<PolicyID>,
    pub errors: Vec<EvaluationError>,
}
```
- Modify the Response struct's 'new' function to accept a `Vec<EvaluationError>` for the errors parameter and update the construction of diagnostics. 
```rust
impl Response {
    pub fn new(decision: Decision, reason: HashSet<PolicyID>, errors: Vec<EvaluationError>) -> Self {
        Response {
            decision,
            diagnostics: Diagnostics { reason, errors },
        }
    }
}

```
The Response struct will now use a `Vec<EvaluationError>` for the errors field within the Diagnostics struct. It's important to 
update any code that creates a new Response instance to pass a `Vec<EvaluationError>` for the errors parameter.
### Final Thoughts 
If changes are satisfactory pull request in 18 hours I need to sleep. All the changes from this thread will be staged: 
(EvaluationError and is_authorized).

**Comment by khieta** [comment]
> Update the function signature to include the return type `Vec<EvaluationError>`

No, the signature of `is_authorized` should stay the same. I just meant to update the `Diagnostics` type, per your comments under the header "Further changes". Sorry for being unclear! 

The approach under the "Further changes" header looks good though. Mark me as a reviewer on your PR.

And make sure to get enough sleep 😆

**Comment by NetwCodeProjects** [comment]
PR Submitted.

**Comment by khieta** [comment]
Note: updated issue title to reflect the new `AuthorizationError` type introduced in #186.

**Comment by khieta** [comment]
Closed by PR #260, which we expect to include in the next major release
